### ISSUE TITLE
Common subexpression elimination

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#687]
This would need some thinking about, but it'd be really nice if we could improve the likes of this:

``` js
var ap = function (__dict_Monad_14) {
    return function (f) {
        return function (a) {
            return $greater$greater$eq(__dict_Monad_14["__superclass_Prelude.Bind_1"]())(f)(function (_2) {
                return $greater$greater$eq(__dict_Monad_14["__superclass_Prelude.Bind_1"]())(a)(function (_1) {
                    return $$return(__dict_Monad_14)(_2(_1));
                });
            });
        };
    };
};
```

into something like this:

``` js
var ap = function (__dict_Monad_14) {
    var __dict_Monad_14_Bind = __dict_Monad_14["__superclass_Prelude.Bind_1"]()
    var $greater$greater$eq_1 = $greater$greater$eq(__dict_Monad_14_Bind);
    var $$return_1 = $$return(__dict_Monad_14);
    return function (f) {
        return function (a) {
            return $greater$greater$eq_1(f)(function (_2) {
                return $greater$greater$eq_1(a)(function (_1) {
                    return $$return_1(_2(_1));
                });
            });
        };
    };
};
```

(Essentially declaring all instance related members and superclass instances as local variables as soon as is possible.)

Apart from improving readability somewhat, using the local `$greater$greater$eq_1` over `$greater$greater$eq(__dict_Monad_14["__superclass_Prelude.Bind_1"]())` everywhere should add up to a nice performance boost too. Most of optimisations I worked on for instances were to make constructing instances cheaper as they were the largest overhead, if we can reduce the number of constructions that's even better!

**Comment by paf31** [comment]
Is there any way to identify TCDs in the core AST?

**Comment by garyb** [comment]
Yep, the `IsTypeClassConstructor` from `Meta` is added to the value for the "class constructor" function during desugaring: https://github.com/purescript/purescript/blob/7b5cb4b3481d1119509c9a26b4d941919c944e1f/src/Language/PureScript/CoreFn/Desugar.hs#L174-L177

**Comment by garyb** [comment]
This might be a bit clearer actually when the revisited typeclass representation is in.

**Comment by rhendric** [comment]
Is there consensus that doing this just for the special case of typeclass dictionaries, without waiting for any other optimization infrastructure, would be a good idea? I'd be interested in taking a stab at it unless there's an expectation that the landscape is likely to change soon—I see @garyb's comment about a revisited typeclass representation, but that comment's four years old now.

**Comment by garyb** [comment]
I think applying the optimisation to typeclass dictionaries in uncontroversial, but personally I'd be fine with it being used more generally. That perhaps needs some discussion still though. The argument for not doing it is that currently the compile output is very predictable, since it's almost 1:1 with the purs code that is written.

The typeclass representation was attempted and stalled twice due to a bug in typechecking, but the bug was finally fixed fairly recently, so I'll perhaps take another stab at it now!
